snuba/query/processors/custom_function.py

```diff
@@ -83,11 +88,15 @@ def apply_function(expression: Expression) -> Expression:
                     )
                 }
 
-                ret = self.__body.transform(
-                    lambda exp: resolved_params[exp.column_name]
-                    if isinstance(exp, Column) and exp.column_name in resolved_params
-                    else exp
-                )
+                def transform_fn(exp: Expression) -> Expression:
+                    if isinstance(exp, Column):
+                        if exp.column_name in resolved_params:
+                            return resolved_params[exp.column_name]
+                        elif exp.column_name in self.__constants:
+                            return Literal(exp.alias, self.__constants[exp.column_name])
+                    return exp
+
+                ret = self.__body.transform(transform_fn)
```

Comment:
It seems we are building partially applied functions where the constants are the parameters you pass when creating the partially applied function itself.
I think we should enforce that you cannot provide the same parameter name in the signature and as a constant.

Also why not doing some refactoring and model this as partially applied function are generally modeled so it would be more familiar and this class will remain as simple as it was before? What about doing something like:
- stop taking a simple string as body.
- add an abstraction (or a function) around the body that parses the string into an AST expression (as the constructor does here). 
- there could be two functions: `simple_function` that just takes the string, parses it and returns the Expression. And `partially_applied` that takes the string, and the constants, parses the string and applies the constants.
- the expression gets passed here to the constructor 
- this functions works as before.
What do you think ?

Reply:
So just so I understand, this would change apdex to be
```
CustomFunction(
    columns,
    "apdex",
    [("column", ColType({UInt})), ("satisfied", LiteralType({int}))],
    simple_function("divide(plus(countIf(lessOrEquals(column, satisfied)), divide(countIf(and(greater(column, satisfied), lessOrEquals(column, multiply(satisfied, 4)))), 2)), count())"),
)
```

and failure rate would be
```
CustomFunction(
    columns,
    "failure_rate",
    [],
    partial_function("divide(countIf(notIn(transaction_status, tuple(ok, cancelled, unknown))), count())", [
        (code, SPAN_STATUS_NAME_TO_CODE[code])
        for code in ("ok", "cancelled", "unknown")
    ]),
)
```

Reply:
I'm not sure that's cleaner/easier to understand. However having a separate parsing function might be useful in other places.

Reply:
My point was not to make it easier for the api user. For them it just looks 99% the same. The goal is to have a clean separation of concerns in the processor itself. Instead of adding another responsibility to the process method in the processor we would have the opportunity to split this out. Then the processor does not deal with parsing anymore and does not care about partially applied functions.

Reply:
OK, I'll split it out. I don't think there's any utility there unless the parsing functions are going to be reused, but it doesn't hurt.